Source/Diagnostics/ReducedDiags/ParticleNumber.cpp

```diff
@@ -0,0 +1,103 @@
+/* Copyright 2019-2020 Neil Zaim, Yinjian Zhao
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "ParticleNumber.H"
+#include "WarpX.H"
+
+// constructor
+ParticleNumber::ParticleNumber (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+    // get WarpX class object
+    auto & warpx = WarpX::GetInstance();
+
+    // get MultiParticleContainer class object
+    auto & mypc = warpx.GetPartContainer();
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
+
+    // resize data array to nSpecies+1
+    // (number of particles of each species + total number of particles)
+    m_data.resize(nSpecies+1, amrex::Real(0.));
+
+    // get species names (std::vector<std::string>)
+    auto species_names = mypc.GetSpeciesNames();
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            ofs << m_sep;
+            ofs << "[3]total()";
+            constexpr int shift_first_species = 4; // Column number of first species in output file
+            for (int i = 0; i < nSpecies; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" + std::to_string(shift_first_species+i) + "]";
+                ofs << species_names[i]+"()";
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes total number of macroparticles
+void ParticleNumber::ComputeDiags (int step)
+{
+
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get MultiParticleContainer class object
+    auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get number of species (int)
+    auto nSpecies = mypc.nSpecies();
```

Comment:
Can this be `const` ?
```suggestion
    const auto nSpecies = mypc.nSpecies();
```